pkg/controller/render/machine_config_to_ignition.go

```diff
@@ -0,0 +1,107 @@
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	"github.com/openshift/machine-config-operator/test/helpers"
+
+	"gopkg.in/yaml.v3"
+)
+
+const MCDContentPath = "/etc/machine-config-daemon/mcd_content.json"
+
+type MCDContent struct {
+	KernelArguments []string `json:"kernelArguments"`
+	FIPS            bool     `json:"fips"`
+}
+
+func MachineConfigToIgnition(mcSpec *mcSpecv1.MachineConfigSpec) (ign3types.Config, error) {
+	// Config
+	ignition, err := ctrlcommon.ParseAndConvertConfig(mcSpec.Config.Raw)
+	if err != nil {
+		return ign3types.Config{}, fmt.Errorf("parsing Ignition config failed: %w", err)
+	}
+
+	var treeFile TreeFile
+	// Extensions
+	supportedExtensions := ctrlcommon.GetSupportedExtensions()
+	for _, mcSpecExtension := range mcSpec.Extensions {
+		treeFile.Packages = append(treeFile.Packages, supportedExtensions[mcSpecExtension]...)
+	}
+
+	// KernelType
+	if ctrlcommon.CanonicalizeKernelType(mcSpec.KernelType) == ctrlcommon.KernelTypeRealtime {
+		treeFile.Packages = append(treeFile.Packages, "kernel-rt", "kernel-rt-modules", "kernel-rt-modules-extra")
+		treeFile.OverrideRemove = append(treeFile.OverrideRemove, "kernel", "kernel-modules", "kernel-modules-extra")
+	}
+
+	// add content from Extensions and KernelType to ignition
+	if !treeFile.isEmpty() {
+		ignitionOriginFile, err := treeFile.toIgnFile(false)
+		if err != nil {
+			return ign3types.Config{}, fmt.Errorf("could not convert OriginFile to Ignition file: %w", err)
+		}
+		ignition.Storage.Files = append(ignition.Storage.Files, ignitionOriginFile)
+	}
+
+	// KernelArguments and FIPS
+	mcdContent, err := json.Marshal(&MCDContent{
+		KernelArguments: mcSpec.KernelArguments,
+		FIPS:            mcSpec.FIPS,
+	})
+	if err != nil {
+		return ign3types.Config{}, fmt.Errorf("couldn't marshal MCD content: %w", err)
+	}
+	ignition.Storage.Files = append(ignition.Storage.Files, helpers.NewIgnFile(MCDContentPath, string(mcdContent)))
+
+	return ignition, nil
+}
+
+type TreeFile struct {
+	Packages       []string `yaml:"packages"`
+	OverrideRemove []string `yaml:"override-remove"`
+}
+
+func (treeFile TreeFile) isEmpty() bool {
+	return len(treeFile.Packages) == 0 && len(treeFile.OverrideRemove) == 0
+}
+
+func (treeFile TreeFile) toIgnFile(allowCompression bool) (ign3types.File, error) {
+	treeFileContents, err := yaml.Marshal(treeFile)
+	if err != nil {
+		return ign3types.File{}, fmt.Errorf("failed to marshal extensions as yaml: %w", err)
+	}
+	fullYamlContents := append([]byte("# Generated by the MCO\n\n"), treeFileContents...)
+	src, gzipped, err := baseutil.MakeDataURL(fullYamlContents, nil, allowCompression)
+	if err != nil {
+		return ign3types.File{}, fmt.Errorf("could not encode extensions file: %w", err)
+	}
+	hash := sha256.New()
+	hash.Write([]byte(src))
+	sha := hex.EncodeToString(hash.Sum(nil))[0:7]
+	mode := 0644
+	file := ign3types.File{
```

Comment:
Hmm it'd clearly be useful to have a "make an ignition file from this data"...oh I see this is a copy of the butane code.  Maybe a public API there?

Anyways, fine for now!

Reply:
(Oh actually we do have `helpers.NewIgnFile`)

Reply:
Yeah and seems like it might even be nice for that to a part f the Ignition API